### ISSUE TITLE
[HttpKernel] `esi` renderer silently falling back to `inline`

### ISSUE THREAD

**Issue by mpdude** [symfony/symfony#35272]
**Symfony version(s) affected**: all

**Description**  

I've always taken this as a matter of fact because it "has been like this" since probably Symfony 2.2. But when I recently talked on the topic, several people were really surprised and asked me to file it as a potential bug.

**Variant 1:**

When using `{{ render_esi(...) }}` and for some reason there is no ESI-capable cache in place (you forgot to enable it, it could be bypassed, ...), the ESI renderer _silently_ falls back to `inline` parsing.

This may be convenient during local development because things "just work":tm:. But it may also catch you by surprise if ESI comes active on deployments in production. For example, you're then suddenly dealing with independent (master) requests where previously subrequests were made. This may have an impact on things present in the `RequestStack`. Also, some kinds of parameters can be passed `inline` but not as real URL parameters in `esi`.

**Variant 2:**

When using `{{ render_esi(...) }}`, the `ResponseCacheStrategy` will make sure to merge `Cache-Control` headers appropriately. When the master response is `Cache-Control: public`, but the subrequest is `Cache-Control: private`, then using ESI correctly will yield a `private` merged response. 

For the `inline` renderer, we just go with the headers set on the master response.

If now, for some reason like above, there is no ESI cache in place, you'll get (`inline`) merging but a `public` response with all the `Cache-Control` settings from the master request – _without any hints_, alarms or warnings that something might be wrong.

**Variant 3:**

It is all too easy to mess up things when you `{{ render(controller(...)) }}` somewhere. Even if care has been taken in the controller to set `Cache-Control` headers correctly, this will use `inline` rendering (be it by lack of awareness/training/understanding, be it by simple mishap to forget the `_esi`) and completely ignore the headers sent in the subrequest.

What do you think – is this something we need/want to change?

Maybe we need not have all the bells and whistles of `ResponseCacheStrategy` when doing `inline` rendering; but would it make sense to fail or warn in cases where `public` and `private` responses get mixed up? (... but what about mixing up Responses with grossly different `max-age` times or similar?)

**Comment by carsonbot** [comment]
Thank you for this issue.
There has not been a lot of activity here for a while. Has this been resolved?

**Comment by mpdude** [comment]
Well, what do others think about that? Is it an issue? How could it be improved?

**Comment by derrabus** [comment]
> When using `{{ render_esi(...) }}` and for some reason there is no ESI-capable cache in place (you forgot to enable it, it could be bypassed, ...), the ESI renderer _silently_ falls back to `inline` parsing.

I think, the main reason for this mechanism is the possibility to develop the application locally without an ESI-capable cache in place. I understand that this might be problematic if the fallback happens unexpectedly on production due to an infrastructure issue.

What would be your suggestion to improve the situation here?

**Comment by fabpot** [comment]
@derrabus gave the reason of the current behavior (which I wouldn't change).

**Comment by mpdude** [comment]
Yes, I completely understand the reason, and unfortunately, I have no good idea right now how this could be improved 😞 (that's why I opened the issue in the first place).

Just for the record, I raised this topic in a stand-up with my team, and one teammate reported that they had struggled with it the other way round recently: For some reason (legacy code...) there was a need to parse and rewrite parts of the HTML/response. Only when the application went into production, it turned out that parts of the response were generated by `{{ render_esi(...) }}` sections. So, suddenly the HTML was missing relevant parts and instead contained ESI tags. Definetely an edge case, but a surprising one.

Feel free to close 😐, since apparently this has been standing for a full year without any clever suggestions.
